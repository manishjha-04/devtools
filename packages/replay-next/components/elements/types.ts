import { ObjectId } from "@replayio/protocol";

export type Attributes = { [key: string]: string };

export type Node = {
  attributes: Attributes;
  children: Node[];
  nodeType: number;
  parentObject: Node | null;
  objectId: ObjectId;
  tagName: string;
};

export type ItemDisplayMode = "collapsed" | "empty" | "head" | "tail";

export type Item = {
  attributes: Attributes;
  depth: number;
  displayMode: ItemDisplayMode;
  displayName: string;
  nodeType: number;
  objectId: ObjectId;
};

export type Metadata = Node & {
  depth: number;
  isExpanded: boolean;
  weight: number;
};
