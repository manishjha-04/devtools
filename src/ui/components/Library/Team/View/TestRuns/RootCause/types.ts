export interface Data {
  discrepancies: {
    failedRun: Run;
    successRun: Run;
    discrepancies: Discrepancy[];
  }[];
}
interface Run {
  id: {
    recordingId: string;
  };
}
export type Discrepancy = ReactComponentDiscrepancyType | ExecutedStatementDiscrepancyType | NetworkEventDiscrepancyType;
export interface ReactComponentDiscrepancyType {
  eventKind: "ReactComponent";
  kind: "Extra" | "Missing";
  sequenceId: string;
  commonPoint: string;
  event: ReactComponentEvent;
}
export interface ExecutedStatementDiscrepancyType {
  eventKind: "ExecutedStatement";
  kind: "Extra" | "Missing";
  sequenceId: string;
  commonPoint: string;
  event: ExecutedStatementEvent;
}
export type NetworkEventDiscrepancyType = __NetworkEventExtraRequestDiscrepancyType | NetworkEventMissingRequestDiscrepancyType | NetworkEventExtraResponseDiscrepancyType | NetworkEventMissingResponseDiscrepancyType;
export interface __NetworkEventExtraRequestDiscrepancyType {
  eventKind: "NetworkEvent";
  kind: "Extra";
  sequenceId: string;
  commonPoint: string;
  event: __NetworkEventExtraRequestEvent;
}
export interface NetworkEventMissingRequestDiscrepancyType {
  eventKind: "NetworkEvent";
  kind: "Missing";
  sequenceId: string;
  commonPoint: string;
  event: NetworkEventMissingRequestEvent;
}
export interface NetworkEventExtraResponseDiscrepancyType {
  eventKind: "NetworkEvent";
  kind: "Extra";
  sequenceId: string;
  commonPoint: string;
  event: NetworkEventExtraResponseEvent;
}
export interface NetworkEventMissingResponseDiscrepancyType {
  eventKind: "NetworkEvent";
  kind: "Missing";
  sequenceId: string;
  commonPoint: string;
  event: NetworkEventMissingResponseEvent;
}
export interface Sequence<T> {
  sequenceId: string;
  kind: "Extra" | "Missing";
  discrepancies: T[];
}
interface ReactComponentEvent extends Event {
  nodeName: string;
}
interface ExecutedStatementEvent extends Event {
  description: {
    column: number;
    line: number;
    text: string;
    url: string;
  }
}
export interface NetworkEventExtraResponseEvent extends Event {
  requestId: string;
  data: NetworkEventResponseEventData;
  alternate: NetworkEventResponseEventData;
}
export interface NetworkEventMissingResponseEvent extends Event {
  requestId: string;
  data: NetworkEventResponseEventData
}
export interface __NetworkEventExtraRequestEvent extends Event {
  requestId: string;
  data: NetworkEventRequestEventData;
}
export interface NetworkEventMissingRequestEvent extends Event {
  requestId: string;
  data: NetworkEventRequestEventData
}
interface Event {
  key: string;
  point: string;
  time: number;
}
interface NetworkEventResponseEventData {
  kind: "ResponseJSON";
  path: string;
  requestTag: string;
  requestUrl: string;
  value: {
    kind: "Any" | "Literal" | "Type" | string;
    type?: "string" | string;
    v?: string[];
  }
}
interface NetworkEventRequestEventData {
  initiator: {
    url: string;
  };
  kind: "Request";
  requestMethod: "GET" | string;
  requestUrl: string;
}