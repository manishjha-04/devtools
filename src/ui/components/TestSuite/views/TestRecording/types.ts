export type Position = "after" | "before";

import { TestEvent, TestSectionName } from "shared/test-suites/RecordingTestMetadata";

export type TestListItem = TestEventWithSectionName | TestSectionEntry;

export interface TestSectionEntry {
  id: string;
  name: TestSectionName;
  title: string;
}

export interface TestSectionEntryWithEvents extends TestSectionEntry {
  events: TestEvent[];
}

export interface TestEventWithSectionName {
  id: string;
  testSectionName: TestSectionName;
  event: TestEvent;
}

export const isTestSectionEntry = (item: TestListItem): item is TestSectionEntry => {
  return "name" in item && "title" in item;
};

export const isTestEventWithName = (item: TestListItem): item is TestEventWithSectionName => {
  return "event" in item;
};

export type TestListDisplayItem = {
  id: string;
  parentId: string | null;
  depth: number;
  item: TestListItem;
  isExpanded: boolean;
};

export type Metadata = {
  id: string;
  parentId: string | null;
  hasChildren: boolean;
  // childrenCanBeRendered: boolean;
  depth: number;
  item: TestListItem;
  isExpanded: boolean;
  subTreeWeight: number;
};
