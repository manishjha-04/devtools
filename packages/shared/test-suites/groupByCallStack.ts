type TreeNode = {
  name: string;
  type: string;
  children?: TreeNode[];
};

type Command = {
  command: string;
  callStack: string[]
}

export function buildCallTree(commands: Command[]): TreeNode | null {
  const root: TreeNode = { name: '', type: 'root', children: [] };

  for (const cmd of commands) {
    const { callStack, command } = cmd

    let currentNode = root;
    for (const func of callStack) {
      let child = currentNode.children?.find(c => c.name === func);
      if (!child) {
        child = { name: func, type: 'function', children: [] };
        currentNode.children?.push(child);
      }
      currentNode = child;
    }
    currentNode.children?.push({ name: command, type: 'command' });
  }

  return root.children && root.children.length > 0 ? root.children[0] : null;
}
