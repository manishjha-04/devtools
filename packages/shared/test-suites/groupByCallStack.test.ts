import { buildCallTree } from "./groupByCallStack";
import { buildCallTree as realBuildCallTree } from "./RecordingTestMetadata";
import events from "./simplified.json"
import eventsWithLines from "./simplified-lines.json"
import eventsWithLines2 from "./simplified-lines-2.json"

import ugly from './events.json'
import ugly2 from './events2.json'

import { TreeNode } from "devtools/client/debugger/src/utils/sources-tree/types";

function simplifyEvents(events: Event[]) {
  return events.map(event => ({
    command: event.command.name,
    callStack: event.call_stack.reverse().map((frame, index) => `${frame.functionName || frame.fileName}.${event.call_stack[index - 1]?.lineNumber || ""}`)
  }))
}

const simplifiedStacks = events.map(e => e.callStack)
const simplifiedStacksWithLines = eventsWithLines.map(e => e.callStack)

function simplifyEvents2(events: any[]) {
  return events.map(event => ({
    command: event.data.command.name,
    callStack: event.data.testSourceCallStack.map(
      (frame, index) => `${frame.functionName || frame.fileName}.${event.data.testSourceCallStack?.[index - 1]?.lineNumber || ""}`
    )
  }))
}




// console.log(JSON.stringify(simplifyEvents2(ugly2), null, 2))

interface Node {
  name: string;
  type: "command" | "function";
  children: Node[];
}

function printNode(node: Node, depth: number = 0) {
  let text = ''
  text += (' '.repeat(depth * 2) + (node.type == "command" ? `_${node.name}_` : node.name.replace(/\.\d*$/, '')) + "\n");

  text += node.children?.map(child => printNode(child, depth + 1))?.join("\n") || "";
  return text;
}

function convertJsonToAscii(root: Node) {
  const res = printNode(root)
  return res.replace(/\n\s*\n/g, '\n').trim();
}

function printRealNode(node: any, depth: number = 0) {
  let text = ''
  text += (' '.repeat(depth * 2) + (node.type == "user-action" ? `_${node.data.command.name}_` : node.data.function.replace(/\.\d*$/, '')) + "\n");

  text += node.data.events?.map(child => printRealNode(child, depth + 1))?.join("\n") || "";
  return text;
}

function convertRealJsonToAscii(root: Node) {
  const res = printRealNode(root)
  return res.replace(/\n\s*\n/g, '\n').trim();
}



function unindent(text: string) {
  const lines = text.split('\n');
  const indent = lines[1].match(/^ */)![0].length;
  return [...lines.slice(1).map(line => line.slice(indent))].join('\n').trim();

}


describe("groupByCallStacks", () => {

  xit("events w/ lines", () => {
    console.log(simplifiedStacksWithLines.length)
    const out = convertJsonToAscii(buildCallTree((eventsWithLines as TreeNode).slice(0, 500)))
    expect(out).toMatchSnapshot()
  })


  xit("raw events 1", () => {
    console.log(JSON.stringify(realBuildCallTree(ugly.events.main.slice(0, 500)), null, 2))
  })

  xit("events 2", () => {
    console.log(convertJsonToAscii(buildCallTree(eventsWithLines2)))
  })

  it("raw events 2", () => {
    console.log(convertRealJsonToAscii(realBuildCallTree(ugly2.slice(0, 5000))))
  })


  xit("add commands", () => {
    function parseInput(input: string) {
      const [cmd, stack] = input.split(":")
      return { command: cmd, callStack: stack.split(",") }
    }
    // Example usage
    const input1 = ["eval: a,b", "goto: a,c"];
    const tree1 = buildCallTree(input1.map(parseInput));
    console.log(JSON.stringify(tree1, null, 2));

    const input2 = ["eval: a,b,c,d", "goto: a,b,e,f", "goto: a,b,e"];
    const tree2 = buildCallTree(input2.map(parseInput));
    console.log(JSON.stringify(tree2, null, 2));
  })
})

