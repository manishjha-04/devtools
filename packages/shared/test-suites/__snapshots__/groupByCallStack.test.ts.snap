// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groupByCallStacks events w/ lines 1`] = `
"tests/breakpoints-01.test.ts
  startTest
    debugPrint
      _page.evaluate_
    _page.goto_
    waitForRecordingToFinishIndexing
      debugPrint
        _page.evaluate_
      waitFor
        callback
          _locator.getAttribute_
          _locator.getAttribute_
  openDevToolsTab
    _locator.isVisible_
    isDevToolsTabActive
      getElementClasses
        _locator.evaluate_
    _locator.click_
    waitFor
      callback
        isDevToolsTabActive
          getElementClasses
            _locator.evaluate_
  addBreakpoint
    openDevToolsTab
      _locator.isVisible_
      isDevToolsTabActive
        getElementClasses
          _locator.evaluate_
    openSource
      _locator.isVisible_
      debugPrint
        _page.evaluate_
      openSourceExplorerPanel
        _locator.isVisible_
      waitFor
        callback
          _locator.count_
      _locator.count_
      _locator.count_
      _locator.click_
      _locator.innerHTML_
      _locator.count_
      _locator.count_
      _locator.click_
      _locator.innerHTML_
      _locator.count_
      _locator.scrollIntoViewIfNeeded_
      _locator.click_
      waitForSelectedSource
        waitFor
          callback
            _locator.getAttribute_
          helpers/source-panel.ts
            _locator.getAttribute_
            mapLocators
              map
                callback
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
                  _locator.textContent_
            _locator.count_
    scrollUntilLineIsVisible
      getSourceLine
        getCurrentSource
          _locator.count_
          _locator.isVisible_
      debugPrint
        _page.evaluate_
      _keyboard.down_
      _keyboard.type_
      _keyboard.up_
      _locator.focus_
      clearTextArea
        debugPrint
          _page.evaluate_
        _locator.focus_
        _keyboard.press_
        _keyboard.press_
      _keyboard.type_
      _keyboard.press_
    waitForSourceLineHitCounts
      getSourceLine
        getCurrentSource
          _locator.count_
          _locator.isVisible_
      _locator.isVisible_
      waitFor
        callback
          _locator.getAttribute_
    getSourceLine
      getCurrentSource
        _locator.count_
        _locator.isVisible_
    waitFor
      callback
        _locator.getAttribute_
    waitFor
      callback
        _mouse.move_
      helpers/source-panel.ts
        _locator.hover_
        _locator.waitFor_
        _locator.getAttribute_
        _locator.click_
    waitForBreakpoint
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
        _locator.click_
      _page.waitForSelector_
      _elementHandle.waitForSelector_
  rewindToLine
    debugPrint
      _page.evaluate_
    clickCommandBarButton
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      _locator.isEnabled_
      _locator.click_
    waitForPaused
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      waitFor
        callback
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
      waitFor
        helpers/pause-information-panel.ts
          getCurrentCallStackFrameInfo
            _locator.waitFor_
            _locator.textContent_
            _locator.textContent_
  executeAndVerifyTerminalExpression
    executeTerminalExpression
      debugPrint
        _page.evaluate_
      openConsolePanel
        _locator.click_
      _locator.waitFor_
      type
        clearText
          focus
            _page.focus_
          _locator.textContent_
        _locator.type_
        submitCurrentText
          _locator.textContent_
          hideTypeAheadSuggestions
            _locator.isVisible_
            _locator.press_
          _locator.textContent_
          _locator.press_
          _locator.textContent_
    verifyEvaluationResult
      debugPrint
        _page.evaluate_
      verifyExpectedCount
        _locator.waitFor_
    clearConsoleEvaluations
      _locator.click_
      waitFor
        callback
          _locator.count_
  rewindToLine
    clickCommandBarButton
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      _locator.isEnabled_
      _locator.click_
    waitForPaused
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      waitFor
        callback
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
      waitFor
        helpers/pause-information-panel.ts
          getCurrentCallStackFrameInfo
            _locator.waitFor_
            _locator.textContent_
            _locator.textContent_
  executeAndVerifyTerminalExpression
    executeTerminalExpression
      debugPrint
        _page.evaluate_
      openConsolePanel
        _locator.click_
      _locator.waitFor_
      type
        clearText
          focus
            _page.focus_
          _locator.textContent_
        _locator.type_
        submitCurrentText
          _locator.textContent_
          hideTypeAheadSuggestions
            _locator.isVisible_
            _locator.press_
          _locator.textContent_
          _locator.press_
          _locator.textContent_
    verifyEvaluationResult
      debugPrint
        _page.evaluate_
      verifyExpectedCount
        _locator.waitFor_
    clearConsoleEvaluations
      _locator.click_
      waitFor
        callback
          _locator.count_
  rewindToLine
    clickCommandBarButton
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      _locator.isEnabled_
      _locator.click_
    waitForPaused
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      waitFor
        callback
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
      waitFor
        helpers/pause-information-panel.ts
          getCurrentCallStackFrameInfo
            _locator.waitFor_
            _locator.textContent_
            _locator.textContent_
  executeAndVerifyTerminalExpression
    executeTerminalExpression
      debugPrint
        _page.evaluate_
      openConsolePanel
        _locator.click_
      _locator.waitFor_
      type
        clearText
          focus
            _page.focus_
          _locator.textContent_
        _locator.type_
        submitCurrentText
          _locator.textContent_
          hideTypeAheadSuggestions
            _locator.isVisible_
            _locator.press_
          _locator.textContent_
          _locator.press_
          _locator.textContent_
    verifyEvaluationResult
      debugPrint
        _page.evaluate_
      verifyExpectedCount
        _locator.waitFor_
    clearConsoleEvaluations
      _locator.click_
      waitFor
        callback
          _locator.count_
  rewindToLine
    clickCommandBarButton
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      _locator.isEnabled_
      _locator.click_
    waitForPaused
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      waitFor
        callback
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
      waitFor
        helpers/pause-information-panel.ts
          getCurrentCallStackFrameInfo
            _locator.waitFor_
            _locator.textContent_
            _locator.textContent_
  executeAndVerifyTerminalExpression
    executeTerminalExpression
      debugPrint
        _page.evaluate_
      openConsolePanel
        _locator.click_
      _locator.waitFor_
      type
        clearText
          focus
            _page.focus_
          _locator.textContent_
        _locator.type_
        submitCurrentText
          _locator.textContent_
          hideTypeAheadSuggestions
            _locator.isVisible_
            _locator.press_
          _locator.textContent_
          _locator.press_
          _locator.textContent_
    verifyEvaluationResult
      debugPrint
        _page.evaluate_
      verifyExpectedCount
        _locator.waitFor_
    clearConsoleEvaluations
      _locator.click_
      waitFor
        callback
          _locator.count_
  rewindToLine
    clickCommandBarButton
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      _locator.isEnabled_
      _locator.click_
    waitForPaused
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      waitFor
        callback
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
      waitFor
        helpers/pause-information-panel.ts
          getCurrentCallStackFrameInfo
            _locator.waitFor_
            _locator.textContent_
            _locator.textContent_
  executeAndVerifyTerminalExpression
    executeTerminalExpression
      debugPrint
        _page.evaluate_
      openConsolePanel
        _locator.click_
      _locator.waitFor_
      type
        clearText
          focus
            _page.focus_
          _locator.textContent_
        _locator.type_
        submitCurrentText
          _locator.textContent_
          hideTypeAheadSuggestions
            _locator.isVisible_
            _locator.press_
          _locator.textContent_
          _locator.press_
          _locator.textContent_
    verifyEvaluationResult
      debugPrint
        _page.evaluate_
      verifyExpectedCount
        _locator.waitFor_
    clearConsoleEvaluations
      _locator.click_
      waitFor
        callback
          _locator.count_
  resumeToLine
    clickCommandBarButton
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      _locator.isEnabled_
      _locator.click_
    waitForPaused
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      waitFor
        callback
          _locator.count_
          _locator.count_
      waitFor
        helpers/pause-information-panel.ts
          getCurrentCallStackFrameInfo
            _locator.waitFor_
            _locator.textContent_
            _locator.textContent_
  executeAndVerifyTerminalExpression
    executeTerminalExpression
      debugPrint
        _page.evaluate_
      openConsolePanel
        _locator.click_
      _locator.waitFor_
      type
        clearText
          focus
            _page.focus_
          _locator.textContent_
        _locator.type_
        submitCurrentText
          _locator.textContent_
          hideTypeAheadSuggestions
            _locator.isVisible_
            _locator.press_
          _locator.textContent_
          _locator.press_
          _locator.textContent_
    verifyEvaluationResult
      debugPrint
        _page.evaluate_
      verifyExpectedCount
        _locator.waitFor_
    clearConsoleEvaluations
      _locator.click_
      waitFor
        callback
          _locator.count_
  resumeToLine
    clickCommandBarButton
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      _locator.isEnabled_
      _locator.click_
    waitForPaused
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      waitFor
        callback
          _locator.count_
          _locator.count_
      waitFor
        helpers/pause-information-panel.ts
          getCurrentCallStackFrameInfo
            _locator.waitFor_
            _locator.textContent_
            _locator.textContent_
  executeAndVerifyTerminalExpression
    executeTerminalExpression
      debugPrint
        _page.evaluate_
      openConsolePanel
        _locator.click_
      _locator.waitFor_
      type
        clearText
          focus
            _page.focus_
          _locator.textContent_
        _locator.type_
        submitCurrentText
          _locator.textContent_
          hideTypeAheadSuggestions
            _locator.isVisible_
            _locator.press_
          _locator.textContent_
          _locator.press_
          _locator.textContent_
    verifyEvaluationResult
      debugPrint
        _page.evaluate_
      verifyExpectedCount
        _locator.waitFor_
    clearConsoleEvaluations
      _locator.click_
      waitFor
        callback
          _locator.count_
  resumeToLine
    clickCommandBarButton
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      _locator.isEnabled_
      _locator.click_
    waitForPaused
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      waitFor
        callback
          _locator.count_
          _locator.count_
      waitFor
        helpers/pause-information-panel.ts
          getCurrentCallStackFrameInfo
            _locator.waitFor_
            _locator.textContent_
            _locator.textContent_
  executeAndVerifyTerminalExpression
    executeTerminalExpression
      debugPrint
        _page.evaluate_
      openConsolePanel
        _locator.click_
      _locator.waitFor_
      type
        clearText
          focus
            _page.focus_
          _locator.textContent_
        _locator.type_
        submitCurrentText
          _locator.textContent_
          hideTypeAheadSuggestions
            _locator.isVisible_
            _locator.press_
          _locator.textContent_
          _locator.press_
          _locator.textContent_
    verifyEvaluationResult
      debugPrint
        _page.evaluate_
      verifyExpectedCount
        _locator.waitFor_
    clearConsoleEvaluations
      _locator.click_
      waitFor
        callback
          _locator.count_
  resumeToLine
    clickCommandBarButton
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      _locator.isEnabled_
      _locator.click_
    waitForPaused
      debugPrint
        _page.evaluate_
      openPauseInformationPanel
        _locator.isVisible_
      waitFor
        callback
          _locator.count_
          _locator.count_
          _locator.count_
          _locator.count_
      waitFor
        helpers/pause-information-panel.ts
          getCurrentCallStackFrameInfo
            _locator.waitFor_
            _locator.textContent_
            _locator.textContent_
  executeAndVerifyTerminalExpression
    executeTerminalExpression
      debugPrint
        _page.evaluate_
      openConsolePanel
        _locator.click_
      _locator.waitFor_
      type
        clearText
          focus
            _page.focus_
          _locator.textContent_
        _locator.type_
        submitCurrentText
          _locator.textContent_
          hideTypeAheadSuggestions
            _locator.isVisible_
            _locator.press_
          _locator.textContent_
          _locator.press_
          _locator.textContent_
    verifyEvaluationResult
      debugPrint
        _page.evaluate_
      verifyExpectedCount
        _locator.waitFor_
    clearConsoleEvaluations
      _locator.click_
      waitFor
        callback
          _locator.count_"
`;
